harness/tests/storage/util.py

```diff
@@ -64,3 +66,94 @@ def run_storage_lifecycle_test(
+        # SharedFS has absolute paths
+        expected_files = {k.lstrip("/"): v for k, v in expected_files.items()}
+
+        full_paths = list_files(local_sync_dir)
+        local_files = [os.path.relpath(fp, storage_relpath) for fp in full_paths]
+        for local_file in local_files:
+            expected_content = expected_files.get(local_file, None)
+            assert expected_content is not None
+
+            with open(os.path.join(storage_relpath, local_file), "rb") as f:
+                observed_content = f.read()
+                if not expected_content == observed_content:
+                    raise AssertionError(
+                        "Expected: '{!r}', Observed: '{!r}'".format(
+                            expected_content, observed_content
+                        )
+                    )
+                expected_files.pop(local_file)
+
+        missing_files = len(expected_files)
+        if missing_files != 0:
+            raise AssertionError(
+                f"There were {missing_files} missing files: {expected_files.keys()}"
+            )
+
+    def timed_backoff_sync_check(expected_files: Dict[str, bytes]) -> None:
+        # Prevent failures due storage propagation delay.
+        num_retries = 5
+        for retry_num in range(num_retries + 1):
+            try:
+                fetcher.fetch_new()
+                verify_files(expected_files)
+                break
+            except AssertionError as e:
+                if retry_num == num_retries:
+                    raise
+                sleep_time = 2 ** retry_num
+                print(f"{e}: sleeping: {sleep_time} seconds")
+                time.sleep(sleep_time)
+
+    files_to_remove = []  # type: List[str]
+
+    try:
+        # (Empty Sync) Ensure empty sync is ok
+        fetcher.fetch_new()
+        local_file_list = list_files(local_sync_dir)
+        assert len(local_file_list) == 0
+
+        # (New Sync) Upload a set of files, sync, ensure files downloaded.
+        first_files = get_filepath_dict([("foo", b"foo"), ("bar", b"bar")])
+        put_files(first_files)
+        files_to_remove.extend(first_files.keys())
+
+        timed_backoff_sync_check(first_files)
+        time.sleep(1)  # racy storage propegation with Azure
+
+        # (Update Sync) Upload new content to same set of files, sync, ensure local files updated.
+        second_files = get_filepath_dict([("foo", b"foo2"), ("bar", b"bar2")])
+        put_files(second_files)
+
+        # Prevent failures due storage propagation delay.
+        timed_backoff_sync_check(second_files)
+
+        # (Additional Sync) Upload a new set of files, sync, ensure new files downloaded.
+        third_files = get_filepath_dict([("baz", b"baz"), ("qux", b"qux")])
+        put_files(third_files)
+        files_to_remove.extend(third_files.keys())
+
+        timed_backoff_sync_check({**second_files, **third_files})
+    finally:
+        rm_files(files_to_remove)
```

Comment:
This is a bit of a strange cleanup pattern, and it's not clear that it's bulletproof.

Why not just create a temporary directory, and put all the files you are going to clean up in there, then delete the whole thing in the `finally` block?

Reply:
> This is a bit of a strange cleanup pattern, and it's not clear that it's bulletproof.

This will cause the test to fail if we fail to remove one of the files. Thoughts on this behavior?

I basically need to delete a dynamic list of files when I exit the test context for any reason. I thought about using a context manager(s) here, but it felt strange because the list is updating and several files get replaced, so don't need to be removed. 

(I probably should have used a dict instead of a list though in the event that the save file is appended to the list twice)


> Why not just create a temporary directory, and put all the files you are going to clean up in there, then delete the whole thing in the finally block?

These are files that are present inside an unknown storage backend. I needed to pass in the storage manager specific `put()` and `rm()` so that I didn't have to copy/paste the test for-each storage backend we have and change 2 lines, like the Fetchers.

Reply:
> I basically need to remove a dynamic list of files from an unknown storage when I exit the test context for any reason

Not if you run the test in a temporary directory you create at the top of the test.  Then you just need to `shutil.rmtree()` a single directory.

Reply:
I don't think you are understanding that these `files_to_remove` exist in S3, Azure or GCS. Would you rather me not test against S3, Azure, and GCS? Or maybe I am not understanding what you mean?

Reply:
talked on zoom, `rm_files` is about storage backend not local files, and this seems reasonable to me